api/src/main/java/io/grpc/NameResolver.java

```diff
@@ -360,6 +399,186 @@ public ConfigOrError parseServiceConfig(Map<String, ?> rawServiceConfig) {
+   *
+   * @since 1.21.0
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1770")
+  public static final class Args {
+    private final int defaultPort;
+    private final ProxyDetector proxyDetector;
+    private final SynchronizationContext syncContext;
+    private final ServiceConfigParser serviceConfigParser;
+
+    Args(Integer defaultPort, ProxyDetector proxyDetector,
+        SynchronizationContext syncContext, ServiceConfigParser serviceConfigParser) {
+      this.defaultPort = checkNotNull(defaultPort, "defaultPort not set");
+      this.proxyDetector = checkNotNull(proxyDetector, "proxyDetector not set");
+      this.syncContext = checkNotNull(syncContext, "syncContext not set");
+      this.serviceConfigParser = checkNotNull(serviceConfigParser, "serviceConfigParser not set");
+    }
+
+    /**
+     * The port number used in case the target or the underlying naming system doesn't provide a
+     * port number.
+     *
+     * @since 1.21.0
+     */
+    public int getDefaultPort() {
+      return defaultPort;
+    }
+
+    /**
+     * If the NameResolver wants to support proxy, it should inquire this {@link ProxyDetector}.
+     * See documentation on {@link ProxyDetector} about how proxies work in gRPC.
+     *
+     * @since 1.21.0
+     */
+    public ProxyDetector getProxyDetector() {
+      return proxyDetector;
+    }
+
+    /**
+     * Returns the {@link SynchronizationContext} where {@link #start(Observer)}, {@link #shutdown}
+     * and {@link #refresh} are run from.
+     *
+     * @since 1.21.0
+     */
+    public SynchronizationContext getSynchronizationContext() {
+      return syncContext;
+    }
+
+    /**
+     * Returns the {@link ServiceConfigParser}.
+     *
+     * @since 1.21.0
+     */
+    public ServiceConfigParser getServiceConfigParser() {
+      return serviceConfigParser;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("defaultPort", defaultPort)
+          .add("proxyDetector", proxyDetector)
+          .add("syncContext", syncContext)
+          .add("serviceConfigParser", serviceConfigParser)
+          .toString();
+    }
+
+    /**
+     * Returns a builder with the same initial values as this object.
+     *
+     * @since 1.21.0
+     */
+    public Builder toBuilder() {
+      Builder builder = new Builder();
+      builder.setDefaultPort(defaultPort);
+      builder.setProxyDetector(proxyDetector);
+      builder.setSynchronizationContext(syncContext);
+      builder.setServiceConfigParser(serviceConfigParser);
+      return builder;
+    }
+
+    /**
+     * Creates a new builder.
+     *
+     * @since 1.21.0
+     */
+    public static Builder newBuilder() {
+      return new Builder();
+    }
+
+    /**
+     * Builder for {@link Args}.
+     *
+     * @since 1.21.0
+     */
+    public static final class Builder {
+      private Integer defaultPort;
```

Comment:
We may want to set defaults for multiple of these, like defaultPort and proxyDetector. But this is good at this point.